shared/js/background/trackers-algo.js

```diff
@@ -0,0 +1,250 @@
+                this.surrogateList = this.processSurrogateList(list.data)
+            }
+        })
+    }
+
+    processTrackerList (data) {
+        for (let name in data) {
+            if (data[name].rules) {
+                for (let i in data[name].rules) {
+                    data[name].rules[i].rule = new RegExp(data[name].rules[i].rule, 'ig')
+                }
+            }
+        }
+        return data
+    }
+
+    processEntityList (data) {
+        const processed = {}
+        for (let entity in data) {
+            data[entity].domains.forEach(domain => {
+                processed[domain] = entity
+            })
+        }
+        return processed
+    }
+
+    processSurrogateList (text) {
+        const b64dataheader = 'data:application/javascript;base64,'
+        const surrogateList = {}
+        const splitSurrogateList = text.trim().split('\n\n')
+
+        splitSurrogateList.forEach(sur => {
+            // remove comment lines
+            const lines = sur.split('\n').filter((line) => {
+                return !(/^#.*/).test(line)
+            })
+
+            // remove first line, store it
+            const firstLine = lines.shift()
+
+            // take identifier from first line
+            const pattern = firstLine.split(' ')[0].split('/')[1]
+            const b64surrogate = Buffer.from(lines.join('\n').toString(), 'binary').toString('base64')
+            surrogateList[pattern] = b64dataheader + b64surrogate
+        })
+        return surrogateList
+    }
+
+    getTrackerData (urlToCheck, siteUrl, request, ops) {
+        ops = ops || {}
+
+        if (!this.entityList || !this.trackerList) {
+            throw new Error('tried to detect trackers before rules were loaded')
+        }
+
+        // single object with all of our requeest and site data split and
+        // processed into the correct format for the tracker set/get functions.
+        // This avoids repeat calls to split and util functions.
+        const requestData = {
+            ops: ops,
+            siteUrl: siteUrl,
+            request: request,
+            siteDomain: this.tldjs.parse(siteUrl).domain,
+            siteUrlSplit: this.utils.extractHostFromURL(siteUrl).split('.'),
+            urlToCheck: urlToCheck,
+            urlToCheckDomain: this.tldjs.parse(urlToCheck).domain,
+            urlToCheckSplit: this.utils.extractHostFromURL(urlToCheck).split('.')
+        }
+
+        // finds a tracker definition by iterating over the whole trackerList and finding the matching tracker.
+        const tracker = this.findTracker(requestData)
+
+        if (!tracker) {
+            return null
+        }
+
+        // finds a matching rule by iterating over the rules in tracker.data and sets redirectUrl.
+        const matchedRule = this.findRule(tracker, requestData)
+
+        const redirectUrl = (matchedRule && matchedRule.surrogate) ? this.surrogateList[matchedRule.surrogate] : false
```

Comment:
So I think we actually want to match on more than just the exact surrogate key. Check out this comment here - https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/jd/tds/shared/js/background/surrogates.es6.js#L56-L59

Reply:
Surrogates are defined in the data now so I'll fix it there.